packages/angular/src/migrations/update-16-1-0/extract-standalone-config-from-bootstrap.ts

```diff
@@ -0,0 +1,187 @@
+
+  let bootstrapCallFilePath = project.targets?.build?.options?.main;
+  let bootstrapCallFileContents = tree.read(bootstrapCallFilePath, 'utf-8');
+
+  const ast = tsquery.ast(bootstrapCallFileContents);
+  const importBootstrapNodes = tsquery(ast, IMPORT_BOOTSTRAP_FILE, {
+    visitAllChildren: true,
+  });
+
+  if (
+    importBootstrapNodes.length > 0 &&
+    importBootstrapNodes[0].getText().includes('./bootstrap')
+  ) {
+    bootstrapCallFilePath = joinPathFragments(
+      project.sourceRoot,
+      'bootstrap.ts'
+    );
+    bootstrapCallFileContents = tree.read(bootstrapCallFilePath, 'utf-8');
+  }
+  return { bootstrapCallFilePath, bootstrapCallFileContents };
+}
+
+function getImportTokenMap(bootstrapCallFileContentsAst: SourceFile) {
+  const importTokenMap = new Map<string, string>();
+  const importedTokensNodes = tsquery(
+    bootstrapCallFileContentsAst,
+    'ImportDeclaration > ImportClause',
+    { visitAllChildren: true }
+  );
+
+  for (const node of importedTokensNodes) {
+    importTokenMap.set(node.getText(), node.parent.getText());
+  }
+  return importTokenMap;
+}
+
+function getImportsRequiredForAppConfig(
+  importTokenMap: Map<string, string>,
+  appConfigNode: Node
+) {
+  const importsRequiredForAppConfig = new Set<string>();
+  const checkImportsForTokens = (nodeText: string) => {
+    const keys = importTokenMap.keys();
+    for (const key of keys) {
+      if (key.includes(nodeText))
+        importsRequiredForAppConfig.add(importTokenMap.get(key));
+    }
+  };
+  const visitEachChild = (node: Node) => {
+    node.forEachChild((node) => {
+      const nodeText = node.getText();
+      checkImportsForTokens(nodeText);
+      visitEachChild(node);
+    });
+  };
+  visitEachChild(appConfigNode);
+  return importsRequiredForAppConfig;
+}
+
+function getAppConfigFileContents(
+  importsRequiredForAppConfig: Set<string>,
+  appConfigText: string
+) {
+  const buildAppConfigFileContents = (
+    importStatements: string[],
+    appConfig: string
+  ) => `import { ApplicationConfig } from '@angular/core';${importStatements.join(
+    '\n'
+  )}
+        export const appConfig: ApplicationConfig = ${appConfig}`;
+
+  const appConfigFileContents = buildAppConfigFileContents(
+    Array.from(importsRequiredForAppConfig),
+    appConfigText
+  );
+  return appConfigFileContents;
+}
+
+function getBootstrapCallFileContents(
+  bootstrapCallFileContents: string,
+  appConfigNode: Node,
+  importsRequiredForAppConfig: Set<string>
+) {
+  let newBootstrapCallFileContents = `import { appConfig } from './app/app.config';
+        ${bootstrapCallFileContents.slice(
+          0,
+          appConfigNode.getStart()
+        )}appConfig${bootstrapCallFileContents.slice(appConfigNode.getEnd())}`;
+
+  for (const importStatement of importsRequiredForAppConfig) {
+    newBootstrapCallFileContents = newBootstrapCallFileContents
+      .split(importStatement)
+      .join('');
```

Comment:
Should we use `replace`? Or is there any benefit to doing this instead?

Reply:
We could likely use replace, it just didn't come into my head when i was writing the code